Added changes for #887 - support for multi-shell

BassemNKhalil opened the pull request:
Added changes to o365 spfx project upgrade to support multi-shell (bash|powershell|cmd) - currently for file copy commands.

waldekmastykarz commented:
Nicely done! To complete the solution, would you mind:
- add support for `mkdir` which we use in rule `FN018001_TEAMS_folder`
- add support for `cat` and `rm` which we use in `src/o365/spfx/commands/project/project-upgrade/rules/FileAddRemoveRule.ts`

I believe these are all shell commands we use at the moment. Looking forward to the update 👏

BassemNKhalil commented:
No, not at all - had a quick look and it seems very similar to the others.
I'll make the changes and submit another PR, hopefully sometime this upcoming week.

waldekmastykarz commented:
Awesome! It would be easier if you could push additional commits to this PR. That way, we'd have everything in one place and could merge it as a complete solution. Appreciate your help! 👏

BassemNKhalil commented:
Added multi-shell support for New-Item (PowerShell) and mkdir (bash and cmd)
Remaining: cat

BassemNKhalil commented:
This should be ready for review ☺️ 
CC: @waldekmastykarz @VelinGeorgiev

waldekmastykarz commented:
@BassemNKhalil seems like we're not covering everything with tests. Have you noticed the coverage drop?

BassemNKhalil commented:
Okay, two questions come to mind 😊 :
1- Can I run Circle CI coverage without committing to GitHub?
2- It seems like my code changes are already covered since the changes to rules and project-upgrade were already covered with tests. The logic I added to support multi-shell cannot be tested in the rules as rules are not aware of the shell targeted (cmd, ps or bash). Any ideas? I hate to be the one decreasing the 100% coverage.

waldekmastykarz commented:
You can run tests including coverage locally using `npm t`. It will generate a coverage report that will tell you exactly what's missing.

BassemNKhalil commented:
Okay, this should now be ready for review. Please let me know when you get a chance to review.

VelinGeorgiev commented on src/o365/spfx/commands/project/project-upgrade.ts:
```diff
@@ -232,17 +295,55 @@ class SpfxProjectUpgradeCommand extends Command {
+        return;
+      }
+      //'Add' support for multiple shells
+      if (f.resolution.startsWith('Add')) {
```
@waldekmastykarz , @BassemNKhalil  quick question. 

Should we make that `Add` token/text more prominent something like `##_Add_##` OR [Add] so we would be sure that only rules that have that token/text will be replaced and not other rules resolutions will be impacted. 

For example is someone not knowing about that implementation here adds a rule that has resolution of a kind `Add text at line 24 in file N` then this piece of code will try to find `'[BEFOREPATH]'` for that particular rule and fail.

 Same applies for words like 'copy', 'Remove' even 'install'

waldekmastykarz replied:
We could make it clearer and use something like `add_cmd` to indicate that it will be replace by the relevant command for the given shell. Good point.

VelinGeorgiev replied:
@BassemNKhalil would you be able to change the 
`Add` with `add_cmd`
`Remove` with `remove_cmd`
`copy` with `copy_cmd` in the rules definitions and also to string replacing based on those tokens.

BassemNKhalil replied:
All done - I've also added create_dir_cmd to the mix.

BassemNKhalil commented:
Just noticed this wasn't referenced in the issue itself, adding the reference for #887

BassemNKhalil commented:
@VelinGeorgiev, thank you for the detailed feedback - I''' be going through this through the week so please bear with me and I'll try and provide updates as I go (as well as separate commits, if that's alright).

> ### add new folder in `PowerShell` shell:
I pushed a commit to fix the first issue (PS New-Item Path exception). While -Path parameter is not required as per the spec you provided, -Name parameter seems to append to the current path so I opted to use -Path and -Name instead of -Name alone. I've also added the quotes as you suggested to support spaces in folder names.
 
> ### delete rule in `CMD` shell
Replacing / with \ for PowerShell (and cmd, I assume) is fine. My concern (and this is where you and @waldekmastykarz can pitch in with your knowledge of the codebase) is that I can see - with a quick search for return './ - const paths that are using / not \\. Do we need to address this on a larger scale across a broader set of commands or is it sufficient to be just this one? If it is just the replace for this command, I'll do that but if we need to address all commands with const paths it may need to be split into its own issue.

waldekmastykarz commented:
Thanks for working on it @BassemNKhalil. Regarding path separators, we could make use of the path separator const that is a part of Node.js, but then we'd either need to fix all paths where we construct them dynamically using the const or we settle with `/` and then in the end, before generating the report, see if we need to replace them for Windows using `\`. I'm good with either approach. We'd need to double check if the separator would be specific to OS or to shell, i.e. does PowerShell on mac require `\` or `/` and what about bash on WSL?

BassemNKhalil commented:
> ### xcopy in `cmd` shell
I replaced xcopy with copy - the only downside is that copy would not create the folder structure but we already do that to keep things aligned with different shells.

> ### add file in `cmd`
Removed the round brackets and tested after, all good.

> ### Example with shell option
Added example to help as requested - let me know if it needs more details

TTBOMK, this addresses all comments - except for path separators relating to the delete command.

@waldekmastykarz I believe behaviour is OS related as PS core on Mac behaves like bash in my quick testing but I don't have access to Win\WSL though. I don't think we can replace separators after the file is created since the file contains URLs as well as npm packages using / and it would be hard to tell which ones to replace after the fact.

BassemNKhalil commented:
Assuming we all agree that paths follow the OS and not the shell, I went ahead and leveraged [NodeJS OS](https://nodejs.org/api/os.html) to check the OS platform. Of course, Istanbul was on to me and I needed to add tests to cover that bit (probably my weakest part so please review closely). I used [mock-os](https://www.npmjs.com/package/mock-os) to mock the platform and added a couple of tests for coverage.

BassemNKhalil commented:
@VelinGeorgiev would you prefer something [like this](https://stackoverflow.com/a/30690853/6031622)? This way we don't need mock-os.

BassemNKhalil commented:
@VelinGeorgiev I've removed mock-os and now rely on sinon to mock process.platform - let me know when you or @waldekmastykarz get a chance to review.

waldekmastykarz commented:
We'll have a look at your changes shortly. Thank you!

waldekmastykarz commented:
@BassemNKhalil if we were to use platform-dependent separators then we should use them only for paths and not through a post-replace, because, as you noticed, it could replace too much, breaking things. If everything works fine on Windows, then we could proceed with `/` and see if it there are any edge cases that would require us to take into account `\` as well.

BassemNKhalil commented:
How about we change the rules themselves to use [node path separator](https://nodejs.org/api/path.html#path_path_sep) instead of hardcoding the path with bash-style backslashes? This way, there's no need for post-replace or any of the complications that comes with it. The code below becomes something like `'.'.concat(path.sep, 'config', path.sep, 'tslint.json');`
https://github.com/pnp/office365-cli/blob/a0602e0e2c1e8f08e2e479bc6882f2e8ff47ef61/src/o365/spfx/commands/project/project-upgrade/rules/FN015004_FILE_config_tslint_json.ts#L6

The change itself is simple enough for me to do but I'll need some help to scope the rules that should be changed. What do you think?

waldekmastykarz commented:
It would make perfect sense, and perhaps there are easier ways to do it. In Node.js there is the `path.normalize` method that allows you to normalize paths. Perhaps you could check if it would automatically resolve `/` to `\` on Windows?
If that's not an option, then the alternative would be to use `path.join`, which is more convenient to use than `concat`.

BassemNKhalil commented:
Removed post-replace as the delete file path is now normalised with path.normalize.

waldekmastykarz commented on src/o365/spfx/commands/project/project-upgrade.ts:
```diff
@@ -232,17 +298,56 @@ class SpfxProjectUpgradeCommand extends Command {
+      }
+      //'Add' support for multiple shells
+      if (f.resolution.startsWith('add_cmd')) {
+        const pathStart = f.resolution.indexOf('[BEFOREPATH]') + '[BEFOREPATH]'.length;
```
Let's add type explicitly

waldekmastykarz commented on src/o365/spfx/commands/project/project-upgrade.ts:
```diff
@@ -232,17 +298,56 @@ class SpfxProjectUpgradeCommand extends Command {
+      //'Add' support for multiple shells
+      if (f.resolution.startsWith('add_cmd')) {
+        const pathStart = f.resolution.indexOf('[BEFOREPATH]') + '[BEFOREPATH]'.length;
+        const pathEnd = f.resolution.indexOf('[AFTERPATH]');
```
Let's add type explicitly

waldekmastykarz commented on src/o365/spfx/commands/project/project-upgrade.ts:
```diff
@@ -232,17 +298,56 @@ class SpfxProjectUpgradeCommand extends Command {
+        const pathEnd = f.resolution.indexOf('[AFTERPATH]');
+        const filePath: string = f.resolution.substring(pathStart, pathEnd);
+
+        const contentStart = f.resolution.indexOf('[BEFORECONTENT]') + '[BEFORECONTENT]'.length;
```
Let's add type explicitly

waldekmastykarz commented on src/o365/spfx/commands/project/project-upgrade.ts:
```diff
@@ -232,17 +298,56 @@ class SpfxProjectUpgradeCommand extends Command {
+        const filePath: string = f.resolution.substring(pathStart, pathEnd);
+
+        const contentStart = f.resolution.indexOf('[BEFORECONTENT]') + '[BEFORECONTENT]'.length;
+        const contentEnd = f.resolution.indexOf('[AFTERCONTENT]');
```
Let's add type explicitly

waldekmastykarz commented on src/o365/spfx/commands/project/project-upgrade.ts:
```diff
@@ -690,6 +816,11 @@ ${f.resolution}
       }
+      if (args.options.shell) {
+        if (['bash', 'powershell', 'cmd'].indexOf(args.options.shell) < 0) {
+          return `${args.options.shell} is not a supported shell. Supported package managers are bash, powershell and cmd`;
```
`Supported shells...`

waldekmastykarz commented on src/o365/spfx/commands/project/project-upgrade.ts:
```diff
@@ -749,6 +880,11 @@ ${f.resolution}
     Get instructions to upgrade the current SharePoint Framework project to the
     latest SharePoint Framework version supported by the Office 365 CLI
       ${this.name}
+
```
This sample should also be added to the command's .md file

waldekmastykarz commented:
Merged manually. Thank you!
